client/fs_endpoint.go

```diff
@@ -0,0 +1,921 @@
+			fmt.Errorf("file %q is a directory", req.Path),
+			helper.Int64ToPtr(400), encoder)
+		return
+	}
+
+	// If offsetting from the end subtract from the size
+	if req.Origin == "end" {
+		req.Offset = fileInfo.Size - req.Offset
+		if req.Offset < 0 {
+			req.Offset = 0
+		}
+	}
+
+	frames := make(chan *sframer.StreamFrame, streamFramesBuffer)
+	errCh := make(chan error)
+	var buf bytes.Buffer
+	frameCodec := codec.NewEncoder(&buf, structs.JsonHandle)
+
+	// Create the framer
+	framer := sframer.NewStreamFramer(frames, streamHeartbeatRate, streamBatchWindow, streamFrameSize)
+	framer.Run()
+	defer framer.Destroy()
+
+	// If we aren't following end as soon as we hit EOF
+	var eofCancelCh chan error
+	if !req.Follow {
+		eofCancelCh = make(chan error)
+		close(eofCancelCh)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start streaming
+	go func() {
+		if err := f.streamFile(ctx, req.Offset, req.Path, req.Limit, fs, framer, eofCancelCh); err != nil {
+			select {
+			case errCh <- err:
+			case <-ctx.Done():
+			}
+		}
+
+		framer.Destroy()
+	}()
+
+	// Create a goroutine to detect the remote side closing
+	go func() {
+		for {
+			if _, err := conn.Read(nil); err != nil {
+				if err == io.EOF {
+					cancel()
+					return
+				}
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	var streamErr error
+OUTER:
+	for {
+		select {
+		case streamErr = <-errCh:
+			break OUTER
+		case frame, ok := <-frames:
+			if !ok {
+				break OUTER
+			}
+
+			var resp cstructs.StreamErrWrapper
+			if req.PlainText {
+				resp.Payload = frame.Data
+			} else {
+				if err = frameCodec.Encode(frame); err != nil {
+					streamErr = err
+					break OUTER
+				}
+
+				resp.Payload = buf.Bytes()
+				buf.Reset()
+			}
+
+			if err := encoder.Encode(resp); err != nil {
+				streamErr = err
+				break OUTER
+			}
+		case <-ctx.Done():
+			break OUTER
+		}
+	}
+
+	if streamErr != nil {
+		f.handleStreamResultError(streamErr, helper.Int64ToPtr(500), encoder)
+		return
+	}
+}
+
+// logs is is used to stream a task's logs.
+func (f *FileSystem) logs(conn io.ReadWriteCloser) {
+	defer metrics.MeasureSince([]string{"client", "file_system", "logs"}, time.Now())
+	defer conn.Close()
+
+	// Decode the arguments
+	var req cstructs.FsLogsRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&req); err != nil {
+		f.handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// Check read permissions
+	if aclObj, err := f.c.ResolveToken(req.QueryOptions.AuthToken); err != nil {
+		f.handleStreamResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil {
+		readfs := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityReadFS)
+		logs := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityReadLogs)
+		if !readfs && !logs {
```

Comment:
If the ACL token has readfs perms but not logs perms, shouldn't that return `ErrPermissionDenied` as well, given this is the logs endpoint?

Reply:
We are allowing either since if you have fs permission you can just cat the log files anyways. so it is just sugar.